### PR TITLE
Add diagnostics and handle errors gracefully / Upgrade Cairo.

### PR DESCRIPTION
**Pull request by wawel37**
### Please, during review check if the speed of diagnostics during doc generation is acceptable.

### CODE REVIEW

**Comment by maciektr** [comment]
Hi!

I've managed to run some benchmarks. For `openzeppelin/cairo-contracts` I get:

|  Test run | Time  |
| ------------- | ------------- |
| Without diagnostics  | 366 ms  |
| With diagnostics  | 8.37 s  |
| With diagnostics, without lowering group | 5.48 s |

I tried to run parallel diagnostics warmup (inspired with https://github.com/starkware-libs/cairo/blob/e4136a412b03aff3ba1dec3eb8d8d22356cd67ae/crates/cairo-lang-compiler/src/diagnostics.rs#L220), but it didn't really help to bring the time down at all, resulting in `6.7 s` (without lowering group).

Please note that compilation of the `cairo-contracts`  project takes about `22 s`. 

Given that, the cost of running diagnostics is obviously high, but nothing concerning. We should do that. 

I would opt for trying to expose an API in the compiler repository, that does not require `LoweringGroup`, as adding it to the `ScarbDocDatabase` even if we don't use it seems quite uncalled for.

**Comment by wawel37** [comment]
A [PR](https://github.com/starkware-libs/cairo/pull/6440) was created in Cairo repository, to extend the API of diagnostics, so the LoweringGroup will no longer be required

**Comment by maciektr** [comment]
I've tried hard to find some input which would result in diagnostics, and this caused it
```
            #[starknet::contract]
            pub(crate) mod DualCaseERC20Mock {
                use starknet::ContractAddress;

                component!(path: ERC20Component, storage: erc20, event: ERC20Event);

                #[storage]
                pub struct Storage {
                    #[substorage(v0)]
                    pub erc20: ERC20Component::Storage
                }

                #[event]
                enum Event {
                    #[flat]
                    ERC20Event: ERC20Component::Event
                }
            }
```
Do you think we should have a test case like that? 🤔

**Comment by wawel37** [comment]
> I've tried hard to find some input which would result in diagnostics, and this caused it
> 
> ```
>             #[starknet::contract]
>             pub(crate) mod DualCaseERC20Mock {
>                 use starknet::ContractAddress;
> 
>                 component!(path: ERC20Component, storage: erc20, event: ERC20Event);
> 
>                 #[storage]
>                 pub struct Storage {
>                     #[substorage(v0)]
>                     pub erc20: ERC20Component::Storage
>                 }
> 
>                 #[event]
>                 enum Event {
>                     #[flat]
>                     ERC20Event: ERC20Component::Event
>                 }
>             }
> ```
> 
> Do you think we should have a test case like that? 🤔

Done

**Review by maciektr — APPROVED** [review]
Great stuff! 
Minor comments attached below.

**Review by maciektr — APPROVED** [review]

File: extensions/scarb-doc/tests/diagnostics.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+          #[storage]
+          ^********^
+
+    error: Invalid copy trait implementation, Candidate impl core::starknet::storage::storage_base::FlattenedStorageCopy::<?0> has an unused generic parameter.
+     --> [..]
+          #[storage]
+          ^********^
+
+    error: Compilation failed.
+      error: process did not exit successfully: exit code: 1
```
Comment (maciektr) [inline extensions/scarb-doc/tests/diagnostics.rs:179]:
```suggestion
    error: Compilation failed.
    error: process did not exit successfully: exit code: 1
```

File: extensions/scarb-doc/tests/diagnostics.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+          #[storage]
+          ^********^
+
+    error: Invalid copy trait implementation, Candidate impl core::starknet::storage::storage_base::FlattenedStorageCopy::<?0> has an unused generic parameter.
+     --> [..]
+          #[storage]
+          ^********^
+
+    error: Compilation failed.
+    error: process did not exit successfully: exit code: 1
```
Comment (maciektr) [inline extensions/scarb-doc/tests/diagnostics.rs:179]:
```suggestion
    error: process did not exit successfully: exit code: 1

```

File: extensions/scarb-doc/tests/diagnostics.rs [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+      "#});
```
Comment (maciektr) [inline extensions/scarb-doc/tests/diagnostics.rs:181]:
```suggestion
    error: process did not exit successfully: exit code: 1"#});
```

File: extensions/scarb-doc/tests/diagnostics.rs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+          #[storage]
+          ^********^
+
+    error: Invalid copy trait implementation, Candidate impl core::starknet::storage::storage_base::FlattenedStorageCopy::<?0> has an unused generic parameter.
+     --> [..]
+          #[storage]
+          ^********^
+
+    error: Compilation failed.
+    error: process did not exit successfully: exit code: 1"#});
```
Comment (maciektr) [inline extensions/scarb-doc/tests/diagnostics.rs:179]:
```suggestion
    error: process did not exit successfully: exit code: 1
    "#});
```

**Pull request merged**